test/compilable/test20695.d

```diff
@@ -0,0 +1,35 @@
+// https://issues.dlang.org/show_bug.cgi?id=20695
+struct Bar
+{
+    this(const ref Bar o) {}
+
+    string a;
+    uint b;
+}
+
+struct Bar1
+{
+    @disable this(int a);
+    this(const ref Bar1 o) {}
+
+    string a;
+    uint b;
+}
+
+struct Bar2
+{
+    this(const ref Bar2 o) {}
+    @disable this(T)(T a) {}
+
+    string a;
+    uint b;
+}
+void main ()
+{
+    Bar b = { a: "Hello", b: 42 };
+    Bar c = Bar("Hello", 42);
+
+    Bar1 b1 = { a: "Hello", b: 42 };
+
+    Bar2 b2 = { a: "Hello", b: 42 };
```

Comment:
No I don't think that should work, since default construction does not work.

Reply:
This is the behavior introduced by #11159 : "A struct must now have at least one non-@disabled constructor to be
ineligible for brace initialization." However, default construction was not updated to have the same behavior.

Reply:
Hum, that's annoying. From my experience, what we want with structs is:
a) By default, a constructor that makes sense (we have this);
b) The ability to define our own constructor, and for it to override the default (we have this);
c) The ability to disable construction for a type (we kinda have this)
d) Prevent copy of an object (we have this via `@disable` copy constructor);
e) The ability to initialize structs with literal syntax when default construction is possible (we kinda have this);
f) The ability to introduce the default construction in the face of `b` (we don't have this);
g) Copy constructor not being treated as a constructor in the above cases;

Point `c` is useful for non-trivial construction of complex object, as sometimes you want to prevent users from instantiating the object themselves, which might be necessary to present a `@safe` interface for example;
However, if brace initialization now bypasses this, it doesn't fully work anymore. What I was really missing myself was `g` (and a bit of `f`, but that's a new feature), otherwise everything else made sense.